### PR TITLE
Implement tests for license(s) by taking care of `license_url`

### PR DESCRIPTION
**Pull request by sglavoie**
**TL;DR**: I gave it a try and "fixed" (to confirm) the existing tests for licenses by also taking into account `license_url`. CKAN=>Frictionless was flawed in its implementation and Frictionless=>CKAN required very minor additions.

---

The work has been split into two commits, but it should not be too hard to follow in one go:

* Commit https://github.com/frictionlessdata/frictionless-ckan-mapper/commit/8c4f15feea5d2e965d0dade8f3ed4c119890adff does CKAN to Frictionless
* Commit https://github.com/frictionlessdata/frictionless-ckan-mapper/commit/b8f37af23435ad0513da22acea9068c29f6fff99 does Frictionless to CKAN
* Comments left to explain a bit more this PR: https://github.com/frictionlessdata/frictionless-ckan-mapper/pull/32/files

Also, Travis approves :grin:.

### CODE REVIEW

File: frictionless_ckan_mapper/ckan_to_frictionless.py [hunk 1]
```diff
@@ -150,14 +150,49 @@ def dataset(ckandict):
     # package_show can have license_id and license_title
     # TODO: do we always license_id i.e. can we have license_title w/o
     # license_id?
-    if ('licenses' not in outdict and 'license_id' in outdict):
-        outdict['licenses'] = [{
-            'name': outdict['license_id'],
-        }]
-        if 'license_title' in outdict:
-            outdict['licenses'][0]['title'] = outdict['license_title']
+    # There's no mention in the docs of license_title nor license_url, plus
```
Comment (sglavoie) [inline frictionless_ckan_mapper/ckan_to_frictionless.py:153]: I am referring to this part of the docs: https://docs.ckan.org/en/2.8/api/index.html#ckan.logic.action.create.package_create

And I'm considering the fields `license_title` and `license_url` as they are part of the package shown by @pdelboca: https://github.com/pdelboca/frictionless-ckan-mapper/commit/ee268c1115f9f19e07c11c9247a4a3cda4eeb236

Because they are not there by default, I treat them all as optional fields (including `license_id`, which is optional according to the docs).

File: frictionless_ckan_mapper/ckan_to_frictionless.py [hunk 2]
```diff
@@ -150,14 +150,49 @@ def dataset(ckandict):
...
+    if 'license_url' in outdict:
+        outdict['licenses'][0]['path'] = outdict['license_url']
     outdict.pop('license_id', None)
     outdict.pop('license_title', None)
+    outdict.pop('license_url', None)
+
+    # Check `ckandict` instead of `outdict` because outdict['extras']
+    # was deleted earlier
+    licenses_there = False  # First assume we won't find licenses
+    if ckandict.get('extras'):
```
Comment (sglavoie) [inline frictionless_ckan_mapper/ckan_to_frictionless.py:174]: For this chunk of code (lines 173 to 195), I think you will find it much easier to follow by referring to the accompanying tests in `tests/test_ckan_to_frictionless.py`:

* `test_dataset_license`
* `test_dataset_license_with_licenses_in_extras`

It's still a bit of a mess, but that's JSON. :shrug:

File: frictionless_ckan_mapper/frictionless_to_ckan.py [hunk 3]
```diff
@@ -88,6 +89,7 @@ def package(fddict):
     if 'licenses' in outdict and outdict['licenses']:
         outdict['license_id'] = outdict['licenses'][0].get('name')
         outdict['license_title'] = outdict['licenses'][0].get('title')
+        outdict['license_url'] = outdict['licenses'][0].get('path')
```
Comment (sglavoie) [inline frictionless_ckan_mapper/frictionless_to_ckan.py:92]: Frictionless to CKAN was pretty much working as expected but missing one possible key, `license_url`.

File: tests/test_ckan_to_frictionless.py [hunk 4]
```diff
@@ -169,56 +169,82 @@ def test_unjsonify_all_extra_values(self):
         assert out == exp
 
     def test_dataset_license(self):
+        # No license_title nor license_url
```
Comment (sglavoie) [inline tests/test_ckan_to_frictionless.py:172]: Mentioning because we're not storing `null` values in the Frictionless format (but we are from Frictionless to CKAN).

File: tests/test_ckan_to_frictionless.py [hunk 5]
```diff
@@ -169,56 +169,82 @@ def test_unjsonify_all_extra_values(self):
...
+                                {
+                                    'name': 'cc-by',
+                                    'title': 'Creative Commons Attribution',
+                                    'path': 'http://www.opendefinition.org/licenses/cc-by'
+                                },
+                                {
+                                    'name': 'odc-odbl',
+                                    'title': 'Open Data Commons Open Database License',
+                                    'path': 'https://opendatacommons.org/licenses/odbl/1-0/index.html'
+                                }
```
Comment (sglavoie) [inline tests/test_ckan_to_frictionless.py:222]: That's a bunch of symbols below, but I think it's still readable enough for this PR in that following the data with the indentation is not too painful. We can refactor if desired (I wanted to avoid hiding stuff away in fixtures. It kills the Python magic but I guess it's better that way :smile:).

File: tests/test_ckan_to_frictionless.py [hunk 6]
```diff
@@ -169,56 +169,82 @@ def test_unjsonify_all_extra_values(self):
...
+        # found and use it first in the list, preserving order (here,
+        # it comes from the root with `license_id`, `license_title` and
+        # `license_url`, not from inside "extras"). If it appears again
+        # in the extras (dicts are equal), don't add it again.
+        # Otherwise, we would have the license 'odc-odbl' twice in this case
         exp = {
-            'licenses': [{
-                'name': 'odc-odbl',
-            }]
+            'licenses': [
```
Comment (sglavoie) [inline tests/test_ckan_to_frictionless.py:236]: As the comment above says (lines 229 to 234), there's a missing license in the output because it checks for duplicates. Also, the first to show in the list of dictionaries is the one from retrieving `license_id`, `license_title` and `license_url`: otherwise, the license with name `cc-by` should be first because it's first in the list of licenses.

File: tests/test_frictionless_to_ckan.py [hunk 7]
```diff
@@ -123,24 +129,27 @@ def test_dataset_license(self):
             'licenses': [
                 {
                     'title': 'Open Data Commons Open Database License',
-                    'type': 'odc-pddl'
+                    'name': 'odc-pddl'
```
Comment (sglavoie) [inline tests/test_frictionless_to_ckan.py:132]: Available properties for a given license according to the specs are `name`, `title` and `path`.
* https://specs.frictionlessdata.io/schemas/data-package.json
* https://specs.frictionlessdata.io/data-package/#licenses

File: frictionless_ckan_mapper/ckan_to_frictionless.py (hunk 2 above)
Comment (rufuspollock) [reply frictionless_ckan_mapper/ckan_to_frictionless.py:174]: I don't understand this logic quite @sglavoie why do i do this afterward. The normal logic would be: if i have already unpacked extras and had licenses i just use that. Why am i unpacking this again??

File: frictionless_ckan_mapper/ckan_to_frictionless.py [hunk 8]
```diff
@@ -150,14 +150,49 @@ def dataset(ckandict):
     # package_show can have license_id and license_title
     # TODO: do we always license_id i.e. can we have license_title w/o
     # license_id?
-    if ('licenses' not in outdict and 'license_id' in outdict):
```
Comment (rufuspollock) [inline frictionless_ckan_mapper/ckan_to_frictionless.py:153]: The logic below is different: this test said: only do this if `licenses` not already in outdict. You've removed that test and hence you have to do all the work on licenses again below. This isn't the logic we wanted.

File: frictionless_ckan_mapper/ckan_to_frictionless.py [hunk 9]
```diff
@@ -150,14 +150,49 @@ def dataset(ckandict):
...
-        outdict['licenses'] = [{
-            'name': outdict['license_id'],
-        }]
-        if 'license_title' in outdict:
-            outdict['licenses'][0]['title'] = outdict['license_title']
+    # There's no mention in the docs of license_title nor license_url, plus
+    # license_id is listed as optional in both CKAN 2.8 and 2.9.
+
+    # Looping like this because all those keys are optional.
```
Comment (rufuspollock) [inline frictionless_ckan_mapper/ckan_to_frictionless.py:156]: You should wrap this whole block in an if statement to check we don't have licenses already via extras (in which case we just use that ...

**Comment by rufuspollock** [comment]
@sglavoie first a comment on commit messages: please make them more detailed. Yes you are making improvements but *what* improvements - add a bit of details in bullets below summary if needed.

File: frictionless_ckan_mapper/frictionless_to_ckan.py (hunk 3 above)
Comment (rufuspollock) [reply frictionless_ckan_mapper/frictionless_to_ckan.py:92]: You don't usually need to annotate your own commits - this stuff should be obvious from commit message (and PR summary).

File: tests/test_ckan_to_frictionless.py [hunk 10]
```diff
@@ -169,56 +169,82 @@ def test_unjsonify_all_extra_values(self):
...
             'licenses': [{
                 'name': 'odc-odbl',
-                'title': 'Open Data Commons Open Database License'
             }]
         }
         out = converter.dataset(indict)
         assert out == exp
 
-        # finally what if license*s* already there ...
+        # Remap everything in licenses
```
Comment (rufuspollock) [inline tests/test_ckan_to_frictionless.py:184]: this seems a less clear comment than what was removed ...

Comment (rufuspollock) [reply tests/test_ckan_to_frictionless.py:184]: ok, now i see you've moved this to a separate test ...

File: frictionless_ckan_mapper/ckan_to_frictionless.py (hunk 8 above)
Comment (rufuspollock) [reply frictionless_ckan_mapper/ckan_to_frictionless.py:153]: I guess there could be an improvement over what we had (which was taking existing licenses in extras) which was updating first item in `licenses` array (if already) with stuff at root of ckan dict with values from `license_id`, `license_title` etc. I still think that can then be cleaner.

Comment (sglavoie) [reply frictionless_ckan_mapper/ckan_to_frictionless.py:153]: Yes, I have adapted it slightly because "license_id" is supposedly optional according to the docs (I think that's what you meant?).

Comment (rufuspollock) [reply frictionless_ckan_mapper/ckan_to_frictionless.py:153]: that's a separate point - though i think pretty much license_id is always there but strictly ys it could be optional. the for loop stuff is a good idea ... and the algorithm is then:

1.  Use extras first
2. Updating first item in licenses array (if already there - or create it as empty) with stuff at root of ckan dict i.e. values from license_id, license_title etc.

File: frictionless_ckan_mapper/ckan_to_frictionless.py (hunk 2 above)
Comment (sglavoie) [reply frictionless_ckan_mapper/ckan_to_frictionless.py:174]: I agree. The idea was:

* Add everything from 'extras' first. That means also 'licenses' if it's there.
* After that, outdict['extras'] is actually removed from the returned dict (line 102).
* If we find one of 'license_id', 'license_title', 'license_url', then we need to consider this information for the first license that will appear in 'licenses', so it just resets that list to contain an empty dict for now.
* We still have access to everything that was passed as an argument from 'ckandict', so we make sure to rebuild the list of licenses from it because it may have been cleared if one of 'license_id', 'license_title', 'license_url' was there.
* Even if one of those keys wasn't there, it will loop and see if there are licenses and only add them if they weren't there already. In this case, the licenses will be there already if we don't remove the 'licenses' when finding one of  'license_id', 'license_title', 'license_url'.

File: frictionless_ckan_mapper/ckan_to_frictionless.py (hunk 9 above)
Comment (sglavoie) [reply frictionless_ckan_mapper/ckan_to_frictionless.py:156]: This wouldn't apply if my idea is good enough (from this comment: https://github.com/frictionlessdata/frictionless-ckan-mapper/pull/32/files#r446375072) because `outdict['extras']` do not exist anymore at this point and 'licenses' can potentially be overwritten and repopulated depending on if we have one of 'license_id', 'license_title', 'license_url'.

**Comment by rufuspollock** [comment]
I'm running out of time to examine this tonight but the current algorithm
seems more complex than the one i set out ... in that you reassess the
extras dict again after setting from license_id etc which should not be
necessary.

**Comment by sglavoie** [comment]
It will probably need some refactoring, I agree. I am not saying this is the best approach but I thought it did what was necessary based on the comment from @pdelboca in his commit (https://github.com/pdelboca/frictionless-ckan-mapper/commit/ee268c1115f9f19e07c11c9247a4a3cda4eeb236). Maybe @pdelboca could chime in and confirm if I'm going in the right direction or if I misunderstood the commit he created? :slightly_smiling_face:
